### PR TITLE
Fix FSDP CPU offloading when shards are already on CPU

### LINKED ISSUE #21804: SIGSEGV (Segmentation Fault) when saving checkpoints with FSDP on CPU
### Bug description

When running FSDP on CPU with `state_dict_type="full"`, attempting to save a checkpoint (or retrieve the state dict) results in a segmentation fault (SIGSEGV).

This is caused by `_get_full_state_dict_context` unconditionally setting `offload_to_cpu=True` in `FullStateDictConfig` and `FullOptimStateDictConfig`.

When FSDP is running on CPU, `offload_to_cpu=True` triggers a use-after-free in PyTorch's `FlatParamHandle.to_cpu()`. Specifically, `.to("cpu")` is a no-op for tensors that are already on CPU, meaning no copy is made (unlike when offloading from CUDA). As a result, the flat parameter and all original-parameter views of it continue to alias the padded unsharded storage. When `to_cpu()` subsequently frees this storage, it leads to a use-after-free. The subsequent `.clone()` call in the post-state-dict hook then crashes with a segmentation fault.

The fix is to only set `offload_to_cpu=True` when the FSDP shards actually live on an accelerator (non-CPU device).

### Steps to reproduce
1. Run a model using FSDP strategy on CPU configured with `state_dict_type="full"`.
2. Attempt to save a checkpoint (which triggers `_get_full_state_dict_context` with `offload_to_cpu=True` implicitly).

### Expected behavior
Checkpoint saving should succeed without crashing when FSDP is run on CPU.

### Actual behavior
The process crashes with a Segmentation Fault (SIGSEGV).

### What version are you seeing the problem on?

master, v2.6

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
import torch
from lightning.fabric import Fabric
from lightning.fabric.strategies import FSDPStrategy

# Subclass to bypass Fabric's strict type check for FSDPStrategy on CPU
class FSDPStrategySubclass(FSDPStrategy):
    pass

# Minimal model
class MyModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(10, 10)
    def forward(self, x):
        return self.layer(x)

def main():
    # Run on CPU with FSDP subclass, devices=2
    strategy = FSDPStrategySubclass(state_dict_type="full")
    fabric = Fabric(accelerator="cpu", devices=2, strategy=strategy)
    fabric.launch()

    model = MyModel()
    model = fabric.setup(model)

    # Trigger checkpoint saving, which will call _get_full_state_dict_context
    # and fail when FSDP is on CPU and offload_to_cpu=True is set.
    state = {"model": model}
    try:
        fabric.save("test_checkpoint.ckpt", state)
        print(f"Rank {fabric.global_rank}: SUCCESS: Checkpoint saved without crash!")
    except Exception as e:
        print(f"Rank {fabric.global_rank}: FAILED with exception: {e}")

if __name__ == "__main__":
    main()
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.6.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

This issue was encountered while running macrobenchmarks, specifically a workload simulating Llama 3.1 8B on CPU using PyTorch Lightning.
The benchmark configures `FSDPStrategy` with `state_dict_type="full"` (referred to as `fsdp_full` in the benchmark) to test consolidated checkpoint saving performance to GCS from CPU-based workers.

cc @ethanwharris @justusschock @lantiga

### PR DESCRIPTION
**Pull request by zhixiangli**
## What does this PR do?

Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/21804

Offloading to CPU when FSDP is on CPU triggers a use-after-free in PyTorch's FlatParamHandle.to_cpu(). This change checks if the parameters are on CPU and disables `offload_to_cpu` if they are.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/fabric/strategies/fsdp.py
Comment (Yonghui-Lee) [inline src/lightning/fabric/strategies/fsdp.py:896]: Should we apply the same fix to `_get_sharded_state_dict_context`?

Comment (zhixiangli) [reply src/lightning/fabric/strategies/fsdp.py:896]: No, it works for fsdp-sharded.

**Comment by zhixiangli** [comment]
Hi @deependujha, @tchaton, just a gentle ping whenever you have a chance. Would you mind taking a look at this PR when convenient? It would unblock me from building a clean CPU-only macrobenchmark setup for GCS and also benefit the community by providing a way to use CPU for debugging and testing, which depends on this change. Thanks!

File: src/lightning/fabric/strategies/fsdp.py [hunk 1]
```diff
@@ -911,8 +911,13 @@ def _get_full_state_dict_context(
     from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
     from torch.distributed.fsdp.api import FullOptimStateDictConfig
 
-    state_dict_config = FullStateDictConfig(offload_to_cpu=True, rank0_only=rank0_only)
-    optim_state_dict_config = FullOptimStateDictConfig(offload_to_cpu=True, rank0_only=rank0_only)
+    # TODO: This can be cleaned up once PyTorch Lightning picks up the PyTorch version containing
+    # the fix https://github.com/pytorch/pytorch/pull/188990 as the root cause is in PyTorch.
+    # Offloading to CPU when FSDP is on CPU triggers a use-after-free in PyTorch's FlatParamHandle.to_cpu().
```
Comment (deependujha) [inline src/lightning/fabric/strategies/fsdp.py:916]: great work. When a new torch version is released with linked changes, we can add a conditional for it.

**Review by deependujha — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Nice !
